sdk/core/core/src/az_json_builder.c

```diff
@@ -2,220 +2,627 @@
 // SPDX-License-Identifier: MIT
 
 #include "az_hex_private.h"
+#include "az_json_private.h"
 #include "az_json_string_private.h"
 #include "az_span_private.h"
 #include <az_json.h>
 #include <az_span_internal.h>
 
 #include <_az_cfg.h>
 
+enum
+{
+  // Max size for an already escaped string value (~ half of INT_MAX)
+  _az_MAX_ESCAPED_STRING_SIZE = 1000000000,
+
+  // In the worst case, an ASCII character represented as a single UTF-8 byte could expand 6x when
+  // escaped.
+  // For example: '+' becomes '\u0043'
+  // Escaping surrogate pairs (represented by 3 or 4 UTF-8 bytes) would expand to 12 bytes (which is
+  // still <= 6x).
+  _az_MAX_EXPANSION_FACTOR_WHILE_ESCAPING = 6,
+
+  _az_MAX_UNESCAPED_STRING_SIZE
+  = _az_MAX_ESCAPED_STRING_SIZE / _az_MAX_EXPANSION_FACTOR_WHILE_ESCAPING, // 166_666_666 bytes
+};
+
 static AZ_NODISCARD az_span _get_remaining_span(az_json_builder* json_builder)
 {
-  return az_span_slice_to_end(json_builder->_internal.json, json_builder->_internal.length);
+  _az_PRECONDITION_NOT_NULL(json_builder);
+
+  return az_span_slice_to_end(
+      json_builder->_internal.destination_buffer, json_builder->_internal.bytes_written);
 }
 
-AZ_NODISCARD static az_result az_json_builder_append_str(az_json_builder* self, az_span value)
+// TODO: Make this a precondition?
+static AZ_NODISCARD bool _az_is_appending_value_valid(az_json_builder* json_builder)
 {
-  _az_PRECONDITION_NOT_NULL(self);
+  _az_PRECONDITION_NOT_NULL(json_builder);
+
+  az_json_token_kind kind = json_builder->_internal.token_kind;
 
-  az_span remaining_json = _get_remaining_span(self);
+  if (_az_json_stack_peek(&json_builder->_internal.bit_stack))
+  {
+    // Cannot write a JSON value within an object without a property name first.
+    // That includes writing the start of an object or array without a property name.
```

Comment:
really?
what about
```json
[[[[1,2]]]]
```
that's a valid Json with no propierty name on it

Reply:
A JSON array is not a JSON object. What you have is a nest JSON array, which you rightfully pointed out is valid JSON. The comment is correct.